chainstate/tx-verifier/src/transaction_verifier/input_output_policy/purposes_check.rs

```diff
@@ -0,0 +1,187 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use common::chain::{
+    block::BlockRewardTransactable, signature::Signable, Transaction, TxInput, TxOutput,
+};
+use consensus::ConsensusPoSError;
+use utils::ensure;
+
+use crate::transaction_verifier::input_output_policy::IOPolicyError;
+
+use super::super::error::{ConnectTransactionError, SpendStakeError};
+
+// TODO: avoid collecting by using FallibleIterator
```

Comment:
Please explain why

Reply:
Well, I think you mean "use FallibleIterator to avoid manually collecting to a Result"